text/0000-cargo-run-deps.md

```diff
@@ -0,0 +1,122 @@
+
+In order to do that, we would relax the constraints of `cargo run` to accept a package id describing any package part of `Cargo.lock` (by virtue of the `--package=<pkgid>` command line parameter). The required steps would be:
+
+- if the specified `pkgid` does not match the package id for any workspace member:
+    - use the resolver against the locked manifest to determine direct matching dependency versions. Note that transitional dependencies are excluded.
+    - restrict the resulting set to development and build dependencies only, as the main program would not be able to use `cargo run` anyway once installed.
+- with the resulting package set:
+    - if it is empty, then we fail with an error.
+    - if it has a size of more than one, fail with an error alerting the user that there is ambiguity and that they need to specify an exact version as part of the package id. Providing a list of alternatives would be a nice touch.
+    - if it has a size of exactly one, then proceed.
+- from the resulting package, select either the target specified by the `--bin` option, or the default binary target if none is specified. If we get no match, we fail while alerting the user.
+- trigger compilation of the corresponding target with the host system as the target system. This is important when cross-compiling.
+- execute the resulting compiled binary, passing the trailing arguments of `cargo run` to it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+[rust-lang/rfcs#3028](https://github.com/rust-lang/rfcs/pull/3028) has an overlap with this feature. They are not mutually exclusive and can be implemented mostly independently, but at some point we expect the implementation to require a rework dependending on the whichever is merged first.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Right now, users are required to manually install dependencies to either a global location, or to a dedicated location but then to also manually adjust their `PATH` variable to compensate. In case a global location is used, multiple versions of binaries cannot be installed in parallel.
+
+This RFC keeps binary dependencies built within the workspace's `target/` directory and avoids for them to pollute or overwrite the user global installation.
+
+It also ensures reproducibility of builds, since it avoids a different, potentially incompatible, version of a binary to be picked up by mistake. This is particularly relevant for some tools such as e.g. `mdbook-bib`, which complain if run with a different library version of `mdbook` they were built against as part of the manifest spec.
+
+The design mimicks several of other package managers of related programming languages. See the [prior art](#prior-art) section. As such, it allows newcomers to get a quick and intuitive way to run binaries.
+
```

Comment:
There was some discussion about a potential `cargo task` subcommand to run project-specific binaries in https://internals.rust-lang.org/t/cargos-next-few-years/9321/8 . It might make sense to mention this here since it had the common goal of running some project-specific binaries from the command line.

It might also provide some additional motivation for this PR: Running framework-specific commands from the command line, such as the mentioned `cargo task generate-migrations`. The syntax with this RFC would be a bit longer, but [using cargo aliases](https://doc.rust-lang.org/cargo/reference/config.html#alias) this could become just `cargo generate-migrations`. All of this is not possible with the "artifact dependencies" RFC.